pkg/plugins/manager.go

```diff
@@ -11,160 +12,104 @@ import (
 	if !ok {
-		return nil, errPluginTypeNotFound
+		return ErrInvalidPluginNotFound
 	}
-	return newPluginWrapper(fmt.Sprintf("%s_%s", pluginType, pluginName), clientConfig)
+	m.clients[name] = c
+	return nil
 }
 
-// newPluginWrapper generates a Wrapper that wraps the go-plugin client.
-func newPluginWrapper(clientName string, clientConfig plugin.ClientConfig) (*Wrapper, error) {
-	logger := hclog.New(&hclog.LoggerOptions{
+func pluginClient(name string, p plugin.Plugin, hc plugin.HandshakeConfig, cmd *exec.Cmd) (*plugin.Client, error) {
+	clientConfig := plugin.ClientConfig{
+		Plugins: map[string]plugin.Plugin{
+			name: p,
+		},
+		AllowedProtocols: allowedProtocols,
+		HandshakeConfig:  hc,
+		Cmd:              cmd,
+	}
+	return newPluginClient(name, clientConfig)
+}
+
+func newPluginClient(clientName string, clientConfig plugin.ClientConfig) (*plugin.Client, error) {
+	hl := hclog.New(&hclog.LoggerOptions{
 		Name:   fmt.Sprintf("%s_logger", clientName),
 		Output: os.Stdout,
 		Level:  hclog.Debug,
 	})
-	clientConfig.Logger = logger
-	client := plugin.NewClient(&clientConfig)
-	return &Wrapper{
-		Client: client,
-		Log:    logging.Default(),
-	}, nil
+	l := logging.Default()
+	hcl := NewHClogger(hl, l)
+	clientConfig.Logger = hcl
+	return plugin.NewClient(&clientConfig), nil
 }
 
-// RegisterPlugin is used to register a new plugin type with a plugin group. It can also introduce a new plugin group.
-func (m *Manager) RegisterPlugin(pluginType PluginType, pluginName PluginName, pluginID PluginIdentity) {
-	if m.pluginTypes == nil {
-		m.pluginTypes = make(PluginTypeNameMap)
+// LoadPluginClient loads a Client of type T.
+func (m *Manager[T]) LoadPluginClient(name string) (*T, error) {
+	if m == nil {
+		return nil, ErrUninitializedManager
 	}
-	ptcm, ok := m.pluginTypes[pluginType]
+	c, ok := m.clients[name]
 	if !ok {
-		ptcm := make(pluginTypeConfigMap)
-		m.pluginTypes[pluginType] = ptcm
+		return nil, ErrPluginNameNotFound
 	}
-	ptcm[pluginName] = clientConfig(pluginID)
-}
-
-// clientConfig generates a plugin.ClientConfig struct to be used to configure the Manager's plugins with the
-// corresponding configurations
-func clientConfig(identity PluginIdentity) plugin.ClientConfig {
-	handshakeConfig := identity.HandshakeConfig
-	allowedProtocols := []plugin.Protocol{
-		plugin.ProtocolGRPC,
+	grpcClient, err := c.Client()
+	if err != nil {
+		return nil, err
 	}
-	return plugin.ClientConfig{
-		HandshakeConfig:  handshakeConfig,
-		Cmd:              exec.Command(identity.ExecutableLocation), //nolint:gosec
-		AllowedProtocols: allowedProtocols,
-		// "Plugins": the different types of plugins that the Plugin serves/consumes. Basically a plugin per
-		// communication type. There will be a single plugin which is the GRPC implementation of the Plugin.
-		Plugins: map[string]plugin.Plugin{
-			identity.ImplName: identity.Impl,
-		},
+	stub, err := grpcClient.Dispense(name)
+	if err != nil {
+		return nil, err
 	}
+	cd, _ := stub.(T)
```

Comment:
IIUC we really _know_ the desired interface of the gRPC client: it has to support exactly the calls that lakeFS knows about and can perform.  So I don't understand the genericity: why not just cast to interface LakeFSOperatorsOnPluginer (possibly with a better name) here?

Reply:
Each gRPC client is configured with a specific interface to interact with the gRPC server, e.g. Diff interface, Merge interface, etc'
Each plugin manager will hold the interface type that its corresponding plugin types implement, e.g. `Manager[Diff]` will hold the Delta, Iceberg, and Parquet plugins to perform a diff respectively, all of them implementing the `Diff` interface.